clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -494,6 +495,16 @@ private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedExce
         return time.milliseconds() - begin;
     }
 
+    private long getTimestamp(String topic, Long timestamp) {
+        // If log append times is used for the topic, we overwrite the timestamp to avoid server side re-compression.
+        if (metadata.isUsingLogAppendTime(topic))
```

Comment:
It is not required to be INHERITED_TIMESTAMP, but it is good to be so. 

Like I answered in your other comment, if a topic is using LogAppendTime and a broker receives a message whose timestamp is not INHERITED_TIMESTAMP, it will overwrite it and do the recompression. So the first batch of a new producer might cause recompression on broker side, but after that, no recompression should be needed. I will add some comments so it is more clear.